File: app/src/main/java/com/kontranik/koreader/test/PageSplitter2.java
package com.kontranik.koreader.test;

import android.graphics.Paint;
import android.graphics.Typeface;
import android.os.Build;
import android.text.SpannableStringBuilder;
import android.text.StaticLayout;
import android.text.TextPaint;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.TypedValue;

import androidx.annotation.RequiresApi;

import com.kontranik.koreader.model.MyStyle;
import com.kontranik.koreader.model.Word;

import java.util.ArrayList;
import java.util.List;

@RequiresApi(api = Build.VERSION_CODES.Q)
public class PageSplitter2 {

    public static final String TAG = "PageSplitter2";

    private final int pageWidth;
    private final int pageHeight;
    float textSize;

    private final List<CharSequence> pages = new ArrayList<>();
    private SpannableStringBuilder currentLine = new SpannableStringBuilder();
    private SpannableStringBuilder currentPage = new SpannableStringBuilder();
    private int currentLineHeight;
    private int pageContentHeight;
    private int currentLineWidth;
    private int textLineHeight;

    TextPaint mTextPaint;


    public PageSplitter2(int pageWidth, int pageHeight, TextPaint paint) {
        this.pageWidth = pageWidth;
        this.pageHeight = pageHeight;
        this.textLineHeight = 0;

        mTextPaint = paint;
        textSize = paint.getTextSize();
    }

    public void append(String text, MyStyle style) {
        String[] paragraphs = text.split("\n", -1);
        int i;
        for (i = 0; i < paragraphs.length - 1; i++) {
            appendText(paragraphs[i], style);
            appendNewLine();
        }
        appendText(paragraphs[i], style);
    }

    private void appendText(String text, MyStyle style) {
        String[] words = text.split(" ", -1);
        int i;
        for (i = 0; i < words.length - 1; i++) {
            appendWord(new Word(words[i] + " ", style));
        }
        appendWord( new Word(words[i] , style) );
    }

    public void appendWord(Word word) {

        mTextPaint.setTextSize( textSize * word.getwRelativeTextSize()  );

        Paint.FontMetrics fm = mTextPaint.getFontMetrics();

        float textHeight = fm.descent - fm.ascent;
        int _lineHeight = (int) Math.ceil(fm.bottom - fm.top + fm.leading);
        // int _lineHeight = (int) Math.ceil(mTextPaint.getFontMetrics(null) * lineSpacingMultiplier + lineSpacingExtra);
        int _wordWidth = (int) Math.ceil(mTextPaint.measureText(word.getData().toString()));

/*        StaticLayout tempLayout = new StaticLayout(word.getData(), mTextPaint, 100000, android.text.Layout.Alignment.ALIGN_NORMAL, lineSpacingMultiplier, lineSpacingExtra, false);
        int lineCount = tempLayout.getLineCount();
        _wordWidth = 0;
        for(int i=0 ; i < lineCount ; i++){
            _wordWidth += tempLayout.getLineWidth(i);
        }*/
        //_lineHeight = tempLayout.getHeight();

        textLineHeight = Math.max(textLineHeight, _lineHeight);

        if ( currentLineWidth + _wordWidth >= pageWidth) {
            appendLineToPage(textLineHeight);
            appendNewLine();
        }
        appendTextToLine(word, _wordWidth);
    }

    public void newLine() {
        appendNewLine();
    }

    private void appendNewLine() {
        currentLine.append("\n");
        appendLineToPage(textLineHeight);
    }

    private void checkForPageEnd() {
        if (pageContentHeight + currentLineHeight > pageHeight) {
            newPage();
        }
    }

    private void newPage() {
        pages.add(currentPage);
        Log.d(TAG, "pageHeight: " + pageHeight + ", pageContentHeight: " + pageContentHeight);
        Log.d(TAG, "new Page");
        currentPage = new SpannableStringBuilder();
        pageContentHeight = 0;
    }

    private void appendLineToPage(int textLineHeight) {

        checkForPageEnd();

        Log.d(TAG, currentLine.toString());
        Log.d(TAG, "pageWidth: " + pageWidth + ", currentLineWidth: " + currentLineWidth);

        currentPage.append(currentLine);

        currentLine = new SpannableStringBuilder();
        currentLineHeight = textLineHeight;
        pageContentHeight += currentLineHeight;
        currentLineWidth = 0;
    }

    private void appendTextToLine(Word word, int wordWidth) {
        currentLineHeight = Math.max(currentLineHeight, textLineHeight);
        currentLine.append(word.getData());
        currentLineWidth += wordWidth;
    }

    public List<CharSequence> getPages() {
        List<CharSequence> copyPages = new ArrayList<CharSequence>(pages);
        SpannableStringBuilder lastPage = new SpannableStringBuilder(currentPage);
        if (pageContentHeight + currentLineHeight > pageHeight) {
            copyPages.add(lastPage);
            lastPage = new SpannableStringBuilder();
        }
        lastPage.append(currentLine);
        copyPages.add(lastPage);
        return copyPages;
    }


}

File: app/src/main/java/com/kontranik/koreader/test/PageSplitter2.kt
package com.kontranik.koreader.test

import android.graphics.Typeface
import android.os.Build
import android.text.*
import android.text.style.StyleSpan
import android.util.Log
import androidx.annotation.RequiresApi
import com.kontranik.koreader.model.Cursor
import com.kontranik.koreader.model.MyStyle
import com.kontranik.koreader.model.Word
import java.util.*

@RequiresApi(api = Build.VERSION_CODES.Q)
class PageSplitter2(private val pageWidth: Int, private val pageHeight: Int, paint: TextPaint, private val lineSpacingMultiplier: Float, private val lineSpacingExtra: Float,) {
    var textSize: Float
    private val pages: MutableList<CharSequence> = ArrayList()
    private var currentLine = SpannableStringBuilder()
    private var currentPage = SpannableStringBuilder()

    private var currentLineHeight: Int
    private var pageContentHeight: Int
    private var currentLineWidth: Int
    private var lastTextLineHeight: Int
    var mTextPaint: TextPaint
    var typeFace: Typeface

    fun append(text: String, style: MyStyle) {
        val paragraphs = text.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var i: Int
        i = 0
        while (i < paragraphs.size - 1) {
            appendText(paragraphs[i], style)
            i++
        }
        appendText(paragraphs[i], style)
    }

    private fun appendText(text: String, style: MyStyle) {
        val words = text.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var i: Int
        i = 0
        while (i < words.size - 1) {
            appendWord(Word(words[i] + " ", style))
            i++
        }
        appendWord(Word(words[i], style))
    }

    fun appendWord(word: Word) {

        val staticLayout = StaticLayout(
                word.data,
                mTextPaint,
                100000,
                Layout.Alignment.ALIGN_NORMAL,
                lineSpacingMultiplier,
                lineSpacingExtra,
                true
        )
        val startLineTop = staticLayout.getLineTop(0)
        val endLineBottom = staticLayout.getLineBottom(0)

        val _lineHeight: Int = endLineBottom - startLineTop;
        val _wordWidth: Int = Math.ceil( staticLayout.getLineWidth(0).toDouble() ).toInt();

        //textLineHeight = Math.max(textLineHeight, _lineHeight);
        if (currentLineWidth + _wordWidth >= pageWidth && currentLineHeight > 0) {
            appendLineToPage()
        }
        appendTextToLine(word.data, _wordWidth, _lineHeight)
    }

    fun newLine() {
        appendNewLine()
    }

    private fun appendNewLine() {
        if (currentLineHeight == 0) currentLineHeight = lastTextLineHeight
        currentLine.append("\n")
        appendLineToPage()
    }

    private fun checkForPageEnd() {
        if (pageContentHeight + currentLineHeight >= pageHeight) {
            newPage()
        }
    }

    private fun newPage() {
        pages.add(currentPage)
        Log.d(TAG, "pageHeight: $pageHeight, pageContentHeight: $pageContentHeight")
        Log.d(TAG, "new Page")
        currentPage = SpannableStringBuilder()
        pageContentHeight = 0
    }

    private fun appendLineToPage() {
        checkForPageEnd()
        if (pageContentHeight > 0) {
            currentLine.append("\n")
        }
        currentPage.append(currentLine)
        pageContentHeight += currentLineHeight
        Log.d(TAG, "curlineheith: $currentLineHeight, pageContentHeight: $pageContentHeight :: $currentLine")
        currentLine = SpannableStringBuilder()
        currentLineWidth = 0
        lastTextLineHeight = currentLineHeight
        currentLineHeight = 0
    }

    private fun appendTextToLine(spannableString: SpannableString, wordWidth: Int, wordHeight: Int) {
        currentLineHeight = Math.max(currentLineHeight, wordHeight)
        currentLine.append(spannableString)
        currentLineWidth += wordWidth
    }

    fun getPages(): List<CharSequence> {
        val copyPages: MutableList<CharSequence> = ArrayList(pages)
        var lastPage = SpannableStringBuilder(currentPage)
        if (pageContentHeight + currentLineHeight > pageHeight) {
            copyPages.add(lastPage)
            lastPage = SpannableStringBuilder()
        }
        lastPage.append(currentLine)
        copyPages.add(lastPage)
        return copyPages
    }

    companion object {
        const val TAG = "PageSplitter2"
    }

    init {
        mTextPaint = TextPaint(paint)
        textSize = paint.textSize
        typeFace = Typeface.create(paint.typeface, Typeface.NORMAL)
        currentLineHeight = 0
        currentLineWidth = 0
        pageContentHeight = 0
        lastTextLineHeight = 0
    }
}